Expose `CircuitData` interners and registers to the `qiskit-circuit` crate

raynelfss opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary
Tracked by #13001 
This commit adds small changes to the visibility of the internal data of the `CircuitData` object, which would allow for an easier conversion into a `DAGCircuit` once #13001 is completed.

### Details and comments:
- Make `qargs_interner`, `cargs_interner`, `qubits` and `clbits` attributes visible to the rest of the `qiskit-circuit` crate. This could allow for easier transfer of data when converting from `CircuitData` to `DAGCircuuit`.
- Make `CircuitData::iter()` method return an exact-size iterator. Which would allow us to determine the size of the circuit without needing to collect the iterator.

### Suggestions
Leave any suggestions in a comment here.

raynelfss committed:
Initial: Expose `CircuitData` interners and registers to the `qiskit-circuit` crate.
- Make the `CircuitData` iter method be an exact-size iterator.

kevinhartman reviewed (commented):
I think we do need these changes, but it might make sense to make all of the fields in `CircuitData` public to the crate, since it feels somewhat arbitrary to restrict some but not all in this case.

The `ExactSizeIterator` change is nice.

raynelfss commented:
@kevinhartman I agree with making most of the fields in `CircuitData` publicly visible except for two of them.
- The `data` field: Although it would be useful, exposing it would allow it to be mutable if the `CircuitData` instance happens to be owned or mutably borrowed. I'd rather just expose an iterator that returns an immutable reference (like we do with `iter()`) and/or expose a function that returns an immutable view of the `data` field.
- The `ParameterTable`: this attribute is fully private and should not be exposed anywhere else as far as I know based on @jakelishman's assessment.

Thus, that's why I think having both interners and the registers (`BitData`) be the only attributes to be made public to the crate is the correct approach for now.

mtreinish commented:
What about `global_phase`?

kevinhartman commented:
> exposing it would allow it to be mutable if the CircuitData instance happens to be owned or mutably borrowed.

That's true, but what I'm getting at is having any of the fields public (e.g. the DAG's bits) makes it possible for things within `qiskit-circuit` to modify ` DAGCircuit` in ways that make it invalid. For example, with the current state of this PR, we could remove the DAG's bits even if it contains operations that use them. It seems odd to force the implementor of `circuit_to_dag` (and the rest of this crate) to use some parts of the proper / public `DAGCircuit` (e.g. `DAGCircuit::iter`) but also expose a select few of its internal fields. My thinking is that because what we're exposing here already gives `qiskit-circuit` the power to create bad data, we shouldn't try to pretend that the visibility has been designed to protect anything.

kevinhartman commented:
I totally closed that by accident!! I'm so sorry @raynelfss >.<

raynelfss commented:
Well in that case do you suggest to have functions that return immutable references instead, or functions that can unpack things. I.e. `unpack_qarg()` or `get_qubits()`, these could return immutable references and allow for cloning but never in-place modifications.

jakelishman commented:
Since I was mentioned: my feeling is probably mostly with Kevin here, on the side of don't make things `pub` if you can break data coherence by mutating them (including assigning a new value to the field).  That includes `global_phase`: you can break data coherence if you assign to it without suitably updating the parameter table.

Having immutable-view getters and safe setters is great.  There'll definitely be places where we need to go further than that, and that'll involve thinking about what APIs we should expose that'll be both useful and efficient.

For `CircuitData` in particular, I think it's probably a good idea to not let the `pub` access get out of hand, even if in other places we've fallen more on the side of permissiveness for now - `CircuitData` is way more core, and the data-coherence requirements _much_ harder to get right.

kevinhartman commented:
> Well in that case do you suggest to have functions that return immutable references instead

I would also vote for this option. Since `PackedInstruction` already exposes interner keys, it feels reasonable to me for `CircuitData` to expose immutable getters for the interners and bits etc. ^^

jakelishman commented:
Fwiw, I do think many places will need to be able to query the interners in one way or another for performance, so either special methods, or making sure that there's a way to get at an immutable reference to an interner will be important, I think.

raynelfss commented:
Alright then, I will make those changes promptly later in the day. Thank you for your input, team 🚀

raynelfss committed:
FIx: Expose immutable views of interners, registers and global phase.
- Revert the changes making the interners and registers visible to the crate `qiskit-circuit`.
- Create methods to expose immutable borrowed views of the interners, registers and global_phase to prevent from mutating the DAGCircuit.
- Add `get_qargs` and `get_cargs` to unpack interned qargs ans cargs.
- Other tweaks and fixes.

eliarbel commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -1151,6 +1151,47 @@ impl CircuitData {
+        &self.qargs_interner
+    }
+
+    /// Returns an immutable view of the Interner used for Qargs
```
```suggestion
    /// Returns an immutable view of the Interner used for Cargs
```

raynelfss committed:
Format: Fix typo in comment

Co-authored-by: Eli Arbel <46826214+eliarbel@users.noreply.github.com>

kevinhartman commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -1151,6 +1151,47 @@ impl CircuitData {
     }
 
+    /// Returns an immutable view of the Interner used for Qargs
+    pub(crate) fn view_qargs_interner(&self) -> &IndexedInterner<Vec<Qubit>> {
```
```suggestion
    pub fn qargs_interner(&self) -> &Interner<Vec<Qubit>> {
```

I think all of these methods should have public visibility, since accessing them seems fair game for transpiler passes etc. that would live in other crates.

We should also aim to follow Rust's guidelines on getter method naming:

https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter

In this case, it's better to just name the method `qargs_interner`.

The `view_` prefix I believe made sense in the case of `PackedOperation` since it makes it clear that there's some sort of work happening under the covers to produce the result. I do not believe its choice had anything to do with the mutability of the returned reference—the return type already provides this information to the caller.

jakelishman replied:
Fwiw, `PackedInstruction::params_view` is called that instead of just `params` because the raw `params` field is also a public `Option<Box<...>>`, and `params_view` is a helper to convert that mess of a type into a `&[Param]`.

kevinhartman reviewed (commented):
Just a comment on how these getters are named.

raynelfss committed:
Refactor: Use naming convention for getters.

eliarbel commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -1151,6 +1151,41 @@ impl CircuitData {
+        &self.global_phase
+    }
+
+    /// Returns an immutable view of the Qubit register of the circuit
```
```suggestion
    /// Returns an immutable view of the Qubits registered in the circuit
```

eliarbel commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -1151,6 +1151,41 @@ impl CircuitData {
+        &self.qubits
+    }
+
+    /// Returns an immutable view of the Classical register of the circuit
```
```suggestion
    /// Returns an immutable view of the Classical bits registered in the circuit
```

eliarbel reviewed (dismissed):
LGTM, just 2 minor updates to doc comments please

raynelfss committed:
Docs: Apply suggestions from code review

- Correct incorrect docstrings for `qubits()` and `clbits()`

Co-authored-by: Eli Arbel <46826214+eliarbel@users.noreply.github.com>

kevinhartman reviewed (commented):
Looks good, one other thing we should do here is make the `mod interner` public in `lib.rs` so other crates can actually import the `Interner` trait, which is needed to do anything with the interner.

jakelishman commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -1151,6 +1151,41 @@ impl CircuitData {
+    /// Unpacks from InternerIndex to `[Clbit]`
+    pub fn get_cargs(&self, index: Index) -> &[Clbit] {
+        self.cargs_interner().intern(index).value
+    }
```
We don't necessarily need these two with the interners exposed - the caller can just do it themselves and have access to whatever new methods, etc, get added in the future (I have something in the works).

mtreinish replied:
FWIW, I added similar methods in #12959 see: https://github.com/Qiskit/qiskit/pull/12959/files#diff-9d10dd27b7540092c29eb23e0fe83a90b8661fb1f452178b60f0e8fb4fd54d60R6165-R6168 I think it's nice to have a helper method for this since it'll be a common operation we'll need to do in transpiler passes and it's less typing.

jakelishman replied:
If you guys think they're worthwhile, I'm not particularly opposed to the shorthand helpers.

kevinhartman reviewed (approved):
LGTM. Thanks @raynelfss!
